### ISSUE TITLE
openshift: `error reserving pod name` causing pods to take a long time to come up

### ISSUE THREAD

**Issue by dustymabe** [cri-o/cri-o#2984]
**Description**

Pods are taking a long time to come up because of `FailedCreatePodSandBox` with the text: `error reserving pod name` as is found [here](https://github.com/cri-o/cri-o/blob/26654d373138e9c93083777391d5f9c4932c83a0/internal/lib/container_server.go#L575-L583).

This is happening a lot but here are two snippets of output from `oc describe pod` that show it takes a long time for the container to be able to get past this: 

<details>

```
Events:
  Type     Reason                  Age                  From                           Message
  ----     ------                  ----                 ----                           -------
  Normal   Scheduled               <unknown>            default-scheduler              Successfully assigned openshift-authentication-operator/authentication-operator-54f5d6c8c9-qpmd5 to dmabe-r662f-master-0
  Warning  FailedCreatePodSandBox  17m                  kubelet, dmabe-r662f-master-0  Failed create pod sandbox: rpc error: code = DeadlineExceeded desc = context deadline exceeded
  Warning  FailedCreatePodSandBox  17m                  kubelet, dmabe-r662f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_authentication-operator-54f5d6c8c9-qpmd5_openshift-authentication-operator_45d93718-e90f-4888-bad6-660cf9c37a87_0 for id f0325581e9547f0087b6e1f6bf1e8f15d5c926f7b1d6694582129358ff54dfb8
  Warning  FailedCreatePodSandBox  16m                  kubelet, dmabe-r662f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_authentication-operator-54f5d6c8c9-qpmd5_openshift-authentication-operator_45d93718-e90f-4888-bad6-660cf9c37a87_0 for id 5910bcf2356ba371ea1e52b721c45d87ade6606b80806755b6c57b5d16f5eafc
  Warning  FailedCreatePodSandBox  16m                  kubelet, dmabe-r662f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_authentication-operator-54f5d6c8c9-qpmd5_openshift-authentication-operator_45d93718-e90f-4888-bad6-660cf9c37a87_0 for id 3b4ba22f8384095d4632e5e3c58623a88d8df60d10aa4f00b236c4d39f4a1c62
  Warning  FailedCreatePodSandBox  16m                  kubelet, dmabe-r662f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_authentication-operator-54f5d6c8c9-qpmd5_openshift-authentication-operator_45d93718-e90f-4888-bad6-660cf9c37a87_0 for id 8bd02642b91dbd1a182f78284e768a3a1acc57c472c645c2f43309c901c6d3cb
  Warning  FailedCreatePodSandBox  16m                  kubelet, dmabe-r662f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_authentication-operator-54f5d6c8c9-qpmd5_openshift-authentication-operator_45d93718-e90f-4888-bad6-660cf9c37a87_0 for id 57bfa663005eb10c5a477a5d8d95686aef31865f793387f6b0858ad445e250ae
  Warning  FailedCreatePodSandBox  15m                  kubelet, dmabe-r662f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_authentication-operator-54f5d6c8c9-qpmd5_openshift-authentication-operator_45d93718-e90f-4888-bad6-660cf9c37a87_0 for id ec0554c041314a041827820c7f14b5e9c2f2e080c4f397f95333a5d2fc73e122
  Warning  FailedCreatePodSandBox  15m                  kubelet, dmabe-r662f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_authentication-operator-54f5d6c8c9-qpmd5_openshift-authentication-operator_45d93718-e90f-4888-bad6-660cf9c37a87_0 for id 02a32352063e690961ae966d392ef55879e6bf46a2248a1da33c9838f6d2f913
  Warning  FailedCreatePodSandBox  15m                  kubelet, dmabe-r662f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_authentication-operator-54f5d6c8c9-qpmd5_openshift-authentication-operator_45d93718-e90f-4888-bad6-660cf9c37a87_0 for id ee3327067bb0acc0ce4e36412f499ac235de46bd958776d552683e9793b9ed7c
  Warning  FailedCreatePodSandBox  7m7s (x38 over 15m)  kubelet, dmabe-r662f-master-0  (combined from similar events): Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_authentication-operator-54f5d6c8c9-qpmd5_openshift-authentication-operator_45d93718-e90f-4888-bad6-660cf9c37a87_0 for id db6f1ccb7fc62c2e401e16916cb49e7e4ee819ffbaa3b5a1696c58a1b5df450a









Events:
  Type     Reason                       Age                     From                           Message
  ----     ------                       ----                    ----                           -------
  Warning  FailedCreatePodSandBox       11m                     kubelet, dmabe-5w55f-master-0  Failed create pod sandbox: rpc error: code = DeadlineExceeded desc = context deadline exceeded
  Warning  FailedCreatePodSandBox       11m                     kubelet, dmabe-5w55f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_installer-5-dmabe-5w55f-master-0_openshift-kube-controller-manager_daa7588c-750d-4212-970a-907efd17dd20_0 for id 1e50e6ab64dc99c37ef56905612b81948c539749a2962215cc5eebf6219ce322
  Warning  FailedCreatePodSandBox       10m                     kubelet, dmabe-5w55f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_installer-5-dmabe-5w55f-master-0_openshift-kube-controller-manager_daa7588c-750d-4212-970a-907efd17dd20_0 for id ae291d1d557eeb0665f9eefcbea4594bf40b6fb05e3ed2f21c49f930e5bc27d6
  Warning  FailedCreatePodSandBox       10m                     kubelet, dmabe-5w55f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_installer-5-dmabe-5w55f-master-0_openshift-kube-controller-manager_daa7588c-750d-4212-970a-907efd17dd20_0 for id d1065506625639bd4f118d5a9ef1b0d15b440aaa8e57ff506b061024acbf5d14
  Warning  FailedCreatePodSandBox       10m                     kubelet, dmabe-5w55f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_installer-5-dmabe-5w55f-master-0_openshift-kube-controller-manager_daa7588c-750d-4212-970a-907efd17dd20_0 for id 73fe198d5db5729c64930fe3d4d9a17e4fe750443f42d12afec4d5f717e4d667
  Warning  FailedCreatePodSandBox       10m                     kubelet, dmabe-5w55f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_installer-5-dmabe-5w55f-master-0_openshift-kube-controller-manager_daa7588c-750d-4212-970a-907efd17dd20_0 for id 731ae29d4981b5e4554f1de3fdae690fb79e79666e3af861c6c9423dcc458fbb
  Warning  FailedCreatePodSandBox       10m                     kubelet, dmabe-5w55f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_installer-5-dmabe-5w55f-master-0_openshift-kube-controller-manager_daa7588c-750d-4212-970a-907efd17dd20_0 for id 08dbae36269cb69c62c01ea08dd9b2acc7aa3a54f692abbab0a39fe2566ff18a
  Warning  FailedCreatePodSandBox       9m49s                   kubelet, dmabe-5w55f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_installer-5-dmabe-5w55f-master-0_openshift-kube-controller-manager_daa7588c-750d-4212-970a-907efd17dd20_0 for id 317c87e2f0579ef4a9eaebf9fa693d50073dce9fe7a077c139266317b73d800c
  Warning  FailedCreatePodSandBox       9m35s                   kubelet, dmabe-5w55f-master-0  Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_installer-5-dmabe-5w55f-master-0_openshift-kube-controller-manager_daa7588c-750d-4212-970a-907efd17dd20_0 for id efd3a7a2c6c00ebe84fecb4f8ff889d5235bf58c21c028346dd2ee1c9a911ecd
  Warning  FailedCreatePodSandBox       5m47s (x17 over 9m20s)  kubelet, dmabe-5w55f-master-0  (combined from similar events): Failed create pod sandbox: rpc error: code = Unknown desc = error reserving pod name k8s_installer-5-dmabe-5w55f-master-0_openshift-kube-controller-manager_daa7588c-750d-4212-970a-907efd17dd20_0 for id 77b69d4329ebb3a278dbb5e75f578829a6ac019075794bc592971a1efd133be6
  Normal   StaticPodInstallerCompleted  3m51s                   static-pod-installer           Successfully installed revision 5
```

</details>

**Steps to reproduce the issue:**
1. Run an openshift install using [openshift-installer](https://github.com/openshift/installer). I used libvirt. 
2. Notice a lot of pods spend time in ContainerCreating state

**Describe the results you received:**

Pods spend extra time in ContainerCreating mentioning `error reserving pod name`. 

**Describe the results you expected:**

Pods to continue on quickly

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `crio --version`:**

```
$ crio --version
crio version 1.16.0-2.rhaos4.3.git8b6bfb5.el8
commit: "d9e22cb1047050f6f573cae89955eb020a3a3af9-dirty"

$ rpm -q cri-o
cri-o-1.16.0-2.rhaos4.3.git8b6bfb5.el8.x86_64

$ oc describe clusterversion | grep -P 'Version:|Image:' | tail -n 2
    Image:              registry.svc.ci.openshift.org/ocp/release@sha256:0374f23478a2548d86537abd07b2059b7c41e226725f727dea705bd8a9af70fc
    Version:            4.3.0-0.nightly-2019-11-19-122017
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Local Libvirt

**Comment by rphillips** [comment]
cc @mrunalp @haircommander @umohnani8

**Comment by laurelnaiad** [comment]
FWIW, I just saw this twice on a freshly built cluster, not using openshift, though using some of its pieces and parts. 

The first instance, it was plaguing kube-scheduler, and after a few minutes I bailed and reset/init'ed the cluster. The second time, it was hitting kube-controller-manager, but I found this issue, reporting that it eventually worked itself out, and indeed it eventually came through, and my cluster is running again.

```
bare metal ubuntu 19.10
kubernetes 1.17.0 installed via kubeadm

**Output of `crio --version`:**
crio version 1.17.0-dev
commit: "29e77e1dd5d4d259382c21fd6847703913ba3733"

**Output of `crun --version`:**
crun version 0.11.15-40d1
commit: 40d11ce6496ce7f8e84a3034dc716f7fb53481c8
spec: 1.0.0
+SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
```

I ran for several days with a similar but non-identical setup prior to this, and didn't see this error. I'm not sure what specifically might have brought this on in this second incarnation of the server.

If there is anything I should grab from logs or other components' info that might be useful, please LMK. Cheers.

**Comment by rphillips** [comment]
@laurelnaiad are there any errors in the crio logs?

**Comment by laurelnaiad** [comment]
Why yes! Yes there were and still are errors in the logs. I've reduce their total-per-reboot by fixing issues with flannel and cgroup alignment. Having cleaned up somewhat, the cluster was certainly healthier – I could reboot with little worry that one or more of the pods would go into the "error reserving pod name" loop - but I could see that it was still occurring sporadically and briefly at startup time.

I just tried #3102, and it seems to completely solve the "error reserving pod name" problem. Yay! :)

I have `Error deleting network: failed to Statfs \"/proc/1542/ns/net\": no such file or directory"` messages that I cannot explain, but it seems like the actual issue reported here is addressed by #3102.

**Comment by haircommander** [comment]
We have spent a lot of time debugging this issue. Part of it was due to failing to clean up after ourselves when we failed to create a pod. That was a very slim number of cases and I feel confident we've closed them up.

The standard case is that the pod really took a long time to set up (often in the networking step). this caused the kubelet to retry the request, even though its first request was already in flight. 

We've added a deadline to network set up, which allows a pod to be cleaned up if network setup takes a long time, before the kubelet asks for the pod again, as well as a check for a timeout at the end of a pod create to make sure we cleanup quickly.

we will continue investigating, but it seems we're closing in on a sufficient set of solutions to consider this bug closed.
